File: cs/secretscmd.py
"""`cs secrets ...`, `cs enroll`, `cs revoke`."""
from __future__ import annotations

import os
import shutil
import sys
from pathlib import Path
from typing import Dict, List, Optional

from . import gitutil, paths, ui
from .config import Machine
from .manifest import Manifest
from .secrets import env_file, get as get_backend, parse_dotenv, dump_dotenv


def _mask(v: str) -> str:
    return v[:3] + "…" + v[-2:] if len(v) > 8 else "…"


def init(repo: Path, m: Machine, interactive: bool = True) -> int:
    get_backend(m).init(repo, m, interactive)
    return 0


def status(repo: Path, m: Machine) -> int:
    ui.head(f"secrets backend: {m.secrets_backend}")
    get_backend(m).status(repo, m)
    return 0


def edit(repo: Path, m: Machine, name: str) -> int:
    b = get_backend(m)
    b.edit(repo, name)
    _commit(repo, m, f"secrets: edit {name}")
    return 0


def set_values(repo: Path, m: Machine, name: str, pairs: List[str]) -> int:
    b = get_backend(m)
    values = b.load_env(repo, name)
    for p in pairs:
        if "=" not in p:
            raise SystemExit(f"cs: expected KEY=VALUE, got '{p}'")
        k, v = p.split("=", 1)
        values[k.strip()] = v
    b.write_env(repo, name, values)
    _commit(repo, m, f"secrets: set {len(pairs)} value(s) in {name}")
    ui.ok(f"{name}: {', '.join(p.split('=', 1)[0] for p in pairs)} stored (encrypted)")
    return 0


def unset_values(repo: Path, m: Machine, name: str, keys: List[str]) -> int:
    b = get_backend(m)
    values = b.load_env(repo, name)
    for k in keys:
        values.pop(k, None)
    b.write_env(repo, name, values)
    _commit(repo, m, f"secrets: unset {len(keys)} value(s) in {name}")
    return 0


def get(repo: Path, m: Machine, name: str, key: Optional[str], show: bool) -> int:
    values = get_backend(m).load_env(repo, name)
    if key:
        if key not in values:
            return 1
        print(values[key] if show else _mask(values[key]))
        return 0
    for k, v in values.items():
        print(f"{k}={v if show else _mask(v)}")
    return 0


def pull(repo: Path, m: Machine, man: Manifest, project: str, force: bool) -> int:
    p = man.projects.get(project)
    if not p:
        raise SystemExit(f"cs: unknown project '{project}'")
    values = get_backend(m).load_env(repo, project)
    if not values:
        ui.warn(f"no secrets stored for {project} (cs secrets push {project} / cs secrets set {project} K=V)")
        return 1
    target = p.checkout_root(man.workspace(m)) / ".env"
    text = dump_dotenv(values)
    if target.exists() and target.read_text() != text and not force:
        ui.fail(f"{paths.contract(target)} exists and differs — `cs secrets diff {project}`; use --force to overwrite")
        return 1
    target.write_text(text)
    target.chmod(0o600)
    _check_ignored(p.checkout_root(man.workspace(m)), target)
    ui.ok(f"wrote {paths.contract(target)} ({len(values)} keys)")
    return 0


def push(repo: Path, m: Machine, man: Manifest, project: str) -> int:
    p = man.projects.get(project)
    if not p:
        raise SystemExit(f"cs: unknown project '{project}'")
    src = p.checkout_root(man.workspace(m)) / ".env"
    if not src.exists():
        raise SystemExit(f"cs: {paths.contract(src)} not found")
    values = parse_dotenv(src.read_text())
    get_backend(m).write_env(repo, project, values)
    _commit(repo, m, f"secrets: {project} .env")
    _check_ignored(p.checkout_root(man.workspace(m)), src)
    ui.ok(f"{project}: {len(values)} keys encrypted into {paths.contract(env_file(repo, project))}")
    return 0


def diff(repo: Path, m: Machine, man: Manifest, project: str) -> int:
    p = man.projects.get(project)
    if not p:
        raise SystemExit(f"cs: unknown project '{project}'")
    stored = get_backend(m).load_env(repo, project)
    local_f = p.checkout_root(man.workspace(m)) / ".env"
    local = parse_dotenv(local_f.read_text()) if local_f.exists() else {}
    rows = []
    for k in sorted(set(stored) | set(local)):
        a, b = stored.get(k), local.get(k)
        if a == b:
            continue
        rows.append([k, _mask(a) if a is not None else ui.dim("-"), _mask(b) if b is not None else ui.dim("-")])
    if rows:
        ui.table(rows, header=["key", "stored", "local .env"])
        return 1
    ui.ok("no differences")
    return 0


def _check_ignored(root: Path, f: Path) -> None:
    if gitutil.is_repo(root):
        p = gitutil.run(["check-ignore", "-q", str(f)], root, check=False)
        if p.returncode != 0:
            ui.warn(f"{f.name} is NOT gitignored in {paths.contract(root)} — add it to .gitignore")


def environment(repo: Path, m: Machine, man: Manifest, project: Optional[str], warn_missing: bool = True) -> Dict[str, str]:
    """os.environ + global secrets + project secrets (project wins)."""
    env = dict(os.environ)
    if env.get("CS_SECRETS_LOADED") == "1":
        return env
    b = get_backend(m)
    if b.name != "none" and not b.ready(repo):
        if warn_missing:
            ui.warn("secrets not available on this machine (cs secrets init / cs enroll) — continuing without them")
        return env
    env.update(b.load_env(repo, "global"))
    if project:
        env.update(b.load_env(repo, project))
    env["CS_SECRETS_LOADED"] = "1"
    return env


def exec_(repo: Path, m: Machine, man: Manifest, project: Optional[str], cmd: List[str]) -> int:
    if not cmd:
        raise SystemExit("cs: secrets exec needs a command after --")
    if project is None:
        p = man.project_for_path(Path.cwd(), m)
        project = p.name if p else None
    env = environment(repo, m, man, project)
    exe = shutil.which(cmd[0]) or cmd[0]
    os.execvpe(exe, cmd, env)
    return 0  # unreachable


def enroll(repo: Path, m: Machine, machine: str) -> int:
    from .secrets import sops as S
    pf = S.machine_pub_file(repo, machine)
    if not pf.exists():
        raise SystemExit(f"cs: {paths.contract(pf)} not found — run `cs secrets init` on {machine} and `cs sync` on both sides first")
    pub = pf.read_text().strip()
    recs = S.recipients(repo)
    if pub in recs:
        ui.ok(f"{machine} is already a recipient")
        return 0
    S.write_recipients(repo, recs + [pub])
    n = S.updatekeys(repo)
    gitutil.run(["add", "-A", ".sops.yaml", "secrets"], repo)
    gitutil.commit(repo, f"secrets: enroll {machine}", "cs", f"cs@{m.name}")
    ui.ok(f"enrolled {machine}; re-encrypted {n} file(s). Run `cs sync` here, then `cs sync` on {machine}.")
    return 0


def revoke(repo: Path, m: Machine, machine: str) -> int:
    from .secrets import sops as S
    pf = S.machine_pub_file(repo, machine)
    pub = pf.read_text().strip() if pf.exists() else ""
    recs = [r for r in S.recipients(repo) if r != pub]
    if pub and pub in S.recipients(repo):
        S.write_recipients(repo, recs)
        n = S.updatekeys(repo)
        if pf.exists():
            gitutil.run(["rm", "-q", "-r", str(pf.parent.relative_to(repo))], repo)
        gitutil.run(["add", "-A", ".sops.yaml", "secrets"], repo)
        gitutil.commit(repo, f"secrets: revoke {machine}", "cs", f"cs@{m.name}")
        ui.ok(f"revoked {machine}; re-encrypted {n} file(s)")
    else:
        ui.warn(f"{machine} was not a recipient")
    b = get_backend(m)
    keys = set()
    for f in sorted((repo / "secrets").rglob("*.env")):
        name = "global" if f.name == "global.env" else f.stem
        keys |= set(b.load_env(repo, name))
    if keys:
        ui.warn("that machine could read these — rotate them at the source: " + ", ".join(sorted(keys)))
    return 0


def recovery(repo: Path, m: Machine) -> int:
    """Generate a recovery age key: pub committed as recipient, private printed ONCE."""
    import subprocess, tempfile
    from .secrets import sops as S
    exe = shutil.which("age-keygen") or str(paths.home() / ".local" / "bin" / "age-keygen")
    with tempfile.TemporaryDirectory() as td:
        kf = Path(td) / "recovery.txt"
        subprocess.run([exe, "-o", str(kf)], check=True, capture_output=True)
        text = kf.read_text()
    pub = next(l.split(":", 1)[1].strip() for l in text.splitlines() if l.startswith("# public key:"))
    priv = next(l for l in text.splitlines() if l.startswith("AGE-SECRET-KEY-"))
    pf = S.machine_pub_file(repo, "recovery")
    pf.parent.mkdir(parents=True, exist_ok=True)
    pf.write_text(pub + "\n")
    S.write_recipients(repo, S.recipients(repo) + [pub])
    n = S.updatekeys(repo)
    gitutil.run(["add", "-A", ".sops.yaml", "secrets", "machines/recovery"], repo)
    gitutil.commit(repo, "secrets: recovery recipient", "cs", f"cs@{m.name}")
    ui.ok(f"recovery recipient added; re-encrypted {n} file(s)")
    ui.info("")
    ui.info(ui.bold("Store this in your password manager now — it is not saved anywhere else:"))
    print(priv)
    ui.info(ui.dim("On a bare machine: write it to ~/.config/sops/age/keys.txt, run `cs secrets init`, enroll the machine's own key, delete it."))
    return 0


def _commit(repo: Path, m: Machine, msg: str) -> None:
    if gitutil.is_repo(repo) and gitutil.is_dirty(repo):
        gitutil.run(["add", "-A", "secrets"], repo)
        gitutil.commit(repo, msg, "cs", f"cs@{m.name}")

File: src/secretscmd.ts
/** cs secrets …, cs trust, cs untrust */
import { chmodSync, existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync, rmSync } from "node:fs";
import { join, relative } from "node:path";
import { spawnSync } from "node:child_process";
import * as git from "./git.js";
import { contract, home } from "./paths.js";
import { which } from "./deps.js";
import { commit, projectForPath, workspace, type Share } from "./share.js";
import { checkoutRoot } from "./checkout.js";
import { dumpDotenv, envFile, getBackend, parseDotenv } from "./secrets/index.js";
import * as S from "./secrets/sops.js";
import * as ui from "./ui.js";

const mask = (v: string) => (v.length > 8 ? v.slice(0, 3) + "…" + v.slice(-2) : "…");
const commitSecrets = (share: Share, msg: string) => commit(share, msg, ["secrets"]);
const checkIgnored = (root: string, f: string) => { if (git.isRepo(root) && git.git(["check-ignore", "-q", f], root, { check: false }).code !== 0) ui.warn(`${relative(root, f)} is NOT gitignored in ${contract(root)} — add it to .gitignore`); };

export async function init(share: Share, interactive = true) { await (await getBackend(share.machine)).init(share, interactive); return 0; }
export async function status(share: Share) { ui.info(ui.bold(`secrets backend: ${share.machine.secretsBackend}`)); (await getBackend(share.machine)).status(share); return 0; }
export async function edit(share: Share, name: string) { (await getBackend(share.machine)).edit(share.path, name); commitSecrets(share, `secrets: edit ${name}`); return 0; }
export async function setValues(share: Share, name: string, pairs: string[]) {
  const repo = share.path; const b = await getBackend(share.machine);
  await ui.spin(`encrypting ${name}…`, async () => { const v = b.loadEnv(repo, name);
    for (const p of pairs) { const i = p.indexOf("="); if (i < 1) throw new Error(`cs: expected KEY=VALUE, got '${p}'`); v[p.slice(0, i).trim()] = p.slice(i + 1); }
    b.writeEnv(repo, name, v); commitSecrets(share, `secrets: set ${pairs.length} value(s) in ${name}`); });
  ui.ok(`${name}: ${pairs.map((p) => p.split("=")[0]).join(", ")} stored (encrypted)`); return 0;
}
export async function unsetValues(share: Share, name: string, keys: string[]) { const repo = share.path; const b = await getBackend(share.machine); const v = b.loadEnv(repo, name); for (const k of keys) delete v[k]; b.writeEnv(repo, name, v); commitSecrets(share, `secrets: unset ${keys.length} value(s) in ${name}`); return 0; }
export async function get(share: Share, name: string, key: string | undefined, show: boolean) {
  const v = (await getBackend(share.machine)).loadEnv(share.path, name);
  if (key) { if (!(key in v)) return 1; console.log(show ? v[key] : mask(v[key])); return 0; }
  for (const [k, val] of Object.entries(v)) console.log(`${k}=${show ? val : mask(val)}`); return 0;
}
export async function pull(share: Share, project: string, force: boolean) {
  const repo = share.path, m = share.machine; const p = share.manifest.projects[project]; if (!p) throw new Error(`cs: unknown project '${project}'`);
  const v = (await getBackend(m)).loadEnv(repo, project); if (!Object.keys(v).length) { ui.warn(`no secrets stored for ${project} (cs secrets push ${project} / cs secrets set ${project} K=V)`); return 1; }
  const root = checkoutRoot(p, workspace(share)), target = join(root, ".env"), text = dumpDotenv(v);
  if (existsSync(target) && readFileSync(target, "utf8") !== text && !force) { ui.fail(`${contract(target)} exists and differs — cs secrets diff ${project}; use --force to overwrite`); return 1; }
  writeFileSync(target, text); chmodSync(target, 0o600); checkIgnored(root, target); ui.ok(`wrote ${contract(target)} (${Object.keys(v).length} keys)`); return 0;
}
export async function push(share: Share, project: string) {
  const repo = share.path, m = share.machine; const p = share.manifest.projects[project]; if (!p) throw new Error(`cs: unknown project '${project}'`);
  const root = checkoutRoot(p, workspace(share)), src = join(root, ".env"); if (!existsSync(src)) throw new Error(`cs: ${contract(src)} not found`);
  const v = parseDotenv(readFileSync(src, "utf8")); (await getBackend(m)).writeEnv(repo, project, v); commitSecrets(share, `secrets: ${project} .env`); checkIgnored(root, src);
  ui.ok(`${project}: ${Object.keys(v).length} keys encrypted into ${contract(envFile(repo, project))}`); return 0;
}
export async function diff(share: Share, project: string) {
  const p = share.manifest.projects[project]; if (!p) throw new Error(`cs: unknown project '${project}'`);
  const stored = (await getBackend(share.machine)).loadEnv(share.path, project); const lf = join(checkoutRoot(p, workspace(share)), ".env"); const local = existsSync(lf) ? parseDotenv(readFileSync(lf, "utf8")) : {};
  const rows = [...new Set([...Object.keys(stored), ...Object.keys(local)])].sort().filter((k) => stored[k] !== local[k]).map((k) => [k, k in stored ? mask(stored[k]) : ui.dim("-"), k in local ? mask(local[k]) : ui.dim("-")]);
  if (rows.length) { ui.table(rows, ["key", "stored", "local .env"]); return 1; } ui.ok("no differences"); return 0;
}
export async function environment(share: Share, project?: string, warnMissing = true): Promise<NodeJS.ProcessEnv> {
  const env = { ...process.env }; if (env.CS_SECRETS_LOADED === "1") return env;
  const repo = share.path; const b = await getBackend(share.machine);
  if (b.name !== "none" && !b.ready(repo)) { if (warnMissing) ui.warn("secrets not available on this machine (cs secrets init / cs trust) — continuing without them"); return env; }
  Object.assign(env, b.loadEnv(repo, "global")); if (project) Object.assign(env, b.loadEnv(repo, project)); env.CS_SECRETS_LOADED = "1"; return env;
}
export async function exec(share: Share, project: string | undefined, cmd: string[]): Promise<number> {
  if (!cmd.length) throw new Error("cs: secrets exec needs a command after --");
  project ??= projectForPath(share, process.cwd())?.name;
  const env = await environment(share, project);
  const p = spawnSync(cmd[0], cmd.slice(1), { stdio: "inherit", env }); return p.status ?? 1;
}
export async function trust(share: Share, machine: string) {
  const repo = share.path; const pf = S.machinePubFile(repo, machine); if (!existsSync(pf)) throw new Error(`cs: ${contract(pf)} not found — run cs secrets init on ${machine} and cs sync on both sides first`);
  const pub = readFileSync(pf, "utf8").trim(); const recs = S.recipients(repo); if (recs.includes(pub)) { ui.ok(`${machine} is already a recipient`); return 0; }
  S.writeRecipients(repo, [...recs, pub]); const n = await ui.spin("re-encrypting secrets for the new recipient…", () => S.updatekeys(repo)); commit(share, `secrets: trust ${machine}`, [".sops.yaml", "secrets"]);
  ui.ok(`${machine} can now decrypt  ${ui.dim(`${n} file(s) re-encrypted`)}`); return 0;
}
export async function untrust(share: Share, machine: string) {
  const repo = share.path, m = share.machine; const pf = S.machinePubFile(repo, machine); const pub = existsSync(pf) ? readFileSync(pf, "utf8").trim() : ""; const recs = S.recipients(repo);
  if (pub && recs.includes(pub)) { S.writeRecipients(repo, recs.filter((r) => r !== pub)); const n = await ui.spin("re-encrypting secrets without that machine…", () => S.updatekeys(repo)); rmSync(join(repo, "machines", machine), { recursive: true, force: true });
    commit(share, `secrets: untrust ${machine}`, [".sops.yaml", "secrets", "machines"]); ui.ok(`untrusted ${machine}; re-encrypted ${n} file(s)`); }
  else ui.warn(`${machine} was not a recipient`);
  const b = await getBackend(m); const keys = new Set<string>();
  for (const name of ["global", ...Object.keys(man_projects(repo))]) for (const k of Object.keys(b.loadEnv(repo, name))) keys.add(k);
  if (keys.size) ui.warn("that machine could read these — rotate them at the source: " + [...keys].sort().join(", ")); return 0;
}
function man_projects(repo: string): Record<string, true> { const d = join(repo, "secrets", "projects"); const out: Record<string, true> = {}; if (existsSync(d)) for (const f of readdirSync(d)) if (f.endsWith(".env")) out[f.slice(0, -4)] = true; return out; }
export async function recovery(share: Share) {
  const repo = share.path; const tmp = join(home(), ".cache", `cs-recovery-${process.pid}.txt`); const exe = which("age-keygen") || join(home(), ".local", "bin", "age-keygen");
  mkdirSync(join(home(), ".cache"), { recursive: true });
  const p = spawnSync(exe, ["-o", tmp], { encoding: "utf8" }); if (p.status !== 0) throw new Error(`cs: age-keygen failed: ${(p.stderr || "").trim()}`);
  const text = readFileSync(tmp, "utf8"); rmSync(tmp, { force: true });
  const pub = text.split("\n").find((l) => l.startsWith("# public key:"))!.split(":")[1].trim(); const priv = text.split("\n").find((l) => l.startsWith("AGE-SECRET-KEY-"))!;
  const pf = S.machinePubFile(repo, "recovery"); mkdirSync(join(repo, "machines", "recovery"), { recursive: true }); writeFileSync(pf, pub + "\n");
  S.writeRecipients(repo, [...S.recipients(repo), pub]); const n = await S.updatekeys(repo); commit(share, "secrets: recovery recipient", [".sops.yaml", "secrets", "machines/recovery"]);
  ui.ok(`recovery recipient added; re-encrypted ${n} file(s)`); ui.note([priv, "", ui.dim("On a bare machine: write it to ~/.config/sops/age/keys.txt, run cs secrets init, trust the machine's own key, delete it.")], "Store this in your password manager now — it is not saved anywhere else");
  return 0;
}

/** Wizard step: make sure this machine can decrypt — trust from another machine, or self-trust with the recovery key. */
export async function ensureRecipient(share: Share, interactive: boolean): Promise<boolean> {
  const repo = share.path, m = share.machine; const b = await getBackend(m); if (b.name === "none" || b.ready(repo)) return true;
  const { existsSync: ex, readdirSync: rd } = await import("node:fs");
  const md = join(repo, "machines");
  const others = ex(md) ? rd(md).filter((d) => d !== m.name && d !== "recovery" && S.recipients(repo).includes((() => { try { return readFileSync(join(md, d, "age.pub"), "utf8").trim(); } catch { return ""; } })())) : [];
  const where = others.length ? `on ${others.map((x) => ui.bold(x)).join(" or ")}` : "on a machine that already has secrets";
  if (!interactive) { ui.warn(`secrets: not a recipient yet — ${where}: cs sync && cs trust ${m.name} && cs sync; then cs sync here`); return false; }
  for (;;) {
    const choice = await ui.select("This machine cannot decrypt secrets yet. How do you want to enable it?", [
      { value: "trust", label: "Trust it from another machine", hint: "recommended — nothing secret is typed or copied" },
      { value: "recovery", label: "Use the recovery key", hint: "paste it once; it is discarded afterwards" },
      { value: "skip", label: "Skip for now", hint: "Claude runs without secrets until then" },
    ]);
    if (choice === "skip") return false;
    if (choice === "recovery") {
      const priv = await ui.password("recovery key (AGE-SECRET-KEY-…)");
      if (!/^AGE-SECRET-KEY-1[A-Z0-9]+$/.test(priv.trim())) { ui.warn("that does not look like an age secret key"); continue; }
      const tmp = join(home(), ".cache", `cs-recovery-${process.pid}.txt`); (await import("node:fs")).mkdirSync(join(home(), ".cache"), { recursive: true });
      writeFileSync(tmp, priv.trim() + "\n", { mode: 0o600 });
      const prev = process.env.SOPS_AGE_KEY_FILE; process.env.SOPS_AGE_KEY_FILE = tmp;
      try {
        const pub = S.publicKey.call(null); // this machine's own public key comes from its own keys file, not the temp one
        const own = readFileSync(S.machinePubFile(repo, m.name), "utf8").trim();
        if (!S.recipients(repo).includes(own)) S.writeRecipients(repo, [...S.recipients(repo), own]);
        const n = await ui.spin("re-encrypting secrets for this machine…", () => S.updatekeys(repo));
        commit(share, `secrets: trust ${m.name} (recovery key)`, [".sops.yaml", "secrets"]);
        ui.ok(`trusted via the recovery key  ${ui.dim(`${n} file(s) re-encrypted`)}`); void pub;
      } catch (e: any) { ui.fail(`could not trust: ${e.message}`); continue; }
      finally { if (prev === undefined) delete process.env.SOPS_AGE_KEY_FILE; else process.env.SOPS_AGE_KEY_FILE = prev; rmSync(tmp, { force: true }); }
      return b.ready(repo);
    }
    ui.note([`${where} run:`, "", `  ${ui.bold(`cs sync && cs trust ${m.name} && cs sync`)}`, "", ui.dim("that machine re-encrypts the secrets so this one can read them — no secret leaves either machine")], "Trust this machine");
    if (!(await ui.proceed("done on the other machine?", "Done — check now", "Skip for now"))) return false;
    const { syncShare } = await import("./sharesync.js");
    await ui.spin("syncing…", () => syncShare(share, { pullOnly: true, timeout: 20 }));
    if (b.ready(repo)) { ui.ok("this machine can decrypt secrets"); return true; }
    ui.warn("still not a recipient — did the other machine run cs sync after trusting it?");
  }
}
